packages/cli/src/utils.ts

```diff
@@ -168,24 +160,61 @@ export async function processNamedFiles(
 }
 
 /**
- * Lists all appmap.json files in a directory, and passes them to a function.
- * With `await`, `listAppMapFiles` blocks until all the files have been processed.
+ * Lists all matching files in a directory, and passes them to an optional function.
  */
-export async function listAppMapFiles(directory: string, fn: (path: string) => Promise<any> | any) {
+export async function findFiles(
+  directory: string,
+  extensionOrPredicate: string | FilePredicate,
+  fn?: ((path: string) => Promise<any> | any) | undefined
+): Promise<string[]> {
   const printDebug = verbose();
   if (printDebug) {
     console.warn(`Scanning ${directory} for AppMaps`);
   }
 
-  const options: IOptions = { debug: printDebug };
-  if (isAbsolute(directory)) {
-    options.cwd = directory;
-  } else {
-    options.cwd = join(process.cwd(), directory);
-  }
-  await Promise.all(
-    (await promisify(glob)('**/*.appmap.json', options)).map((f) => join(directory, f)).map(fn)
-  );
+  const matchFile = (file: string): boolean => {
+    if (typeof extensionOrPredicate === 'string') {
+      return file.endsWith(extensionOrPredicate);
+    } else {
+      return extensionOrPredicate(file);
+    }
+  };
+
+  const traverseDirectory = async (
+    dir: string,
+    result = new Array<string>()
+  ): Promise<string[]> => {
+    let files: string[];
+    try {
+      files = await readdir(dir);
+    } catch (err) {
+      const code = (err as any).code;
+      if (code === 'ENOENT') return result;
```

Comment:
We get a fair amount of `EPERM` in the wild as well. To resolve this, there are cases where we set the glob option `strict: false` to silently ignore permission errors. It doesn't look like any of those cases are changed in this commit though, so this is really more of an FYI.